ONSdigital/rm-actionexporter-service
# Fix export file overwrite bug

# Motivation and Context
We have been seeing the acceptance tests fail quite regularly in the CI pipeline. From investigation it looks like there might be some problems with race conditions, locking and a general design flaw where we had little protection from re-doing work which was already in-flight.

# What has changed
Completely overhauled the locking to use the simplest possible solution - a global lock so that we guarantee that only one thread and process instance is ever writing to the SFTP server.

Added in a `SendState` class for `ActionRequestInstruction` so that we can capture the transition between created, sending in progress, successfully sent and failed. This should guarantee that we never send out any duplicates.

# How to test?
Set up a number of respondents to receive paper notifications and check to make sure the file is being created on the FTP server with all the expected details in it.

Run the acceptance tests. Hopefully the CI pipeline will be more stable and not fail so regularly with the notify file not being there or containing the wrong data.

# Links
Trello: https://trello.com/c/MLoPHcVk/402-bug-check-for-iac-code-in-sftp-file-intermittently-causes-ats-to-fail
